src/test_objects/stat_tester.hh

```diff
@@ -0,0 +1,119 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __STAT_TESTER_HH__
+#define __STAT_TESTER_HH__
+
+#include "base/statistics.hh"
+#include "params/ScalarStatTester.hh"
+#include "params/StatTester.hh"
+#include "sim/sim_object.hh"
+
+namespace gem5
+{
+
+/**
+ * This classes are used to test the stats system from setting through to
+ * output. The value of the stats are set via the parameters and when
+ * included in a simulation will create these stats and update them at Tick
+ * zero.
+ *
+ * Example usage
+ * -------------
+ *
+ *     import m5
+ *     from m5.objects import (
+ *         Root,
+ *         ScalarStatTester,
+ *     )
+ *
+ *     stat_tester = ScalarStatTester()
+ *
+ *     stat_tester.value = 42
+ *     stat_tester.name = "example"
+ *     root = Root(full_system=False, system=stat_tester)
+ *
+ *     m5.instantiate()
+ *     m5.simulate()
+ *
+ * When the above is run, the m5out/stats.txt file will contain the
+ * following:
+ *
+ *
+ *     ---------- Begin Simulation Statistics ----------
+ *     simSeconds                               18446744.073710
+ *     simTicks                                 18446744073709551616
+ *     finalTick                                18446744073709551616
+ *     simFreq                                  1000000000000
+ *     hostSeconds                                      0.00
+ *     hostTickRate                             111124964299455114248192
+ *     hostMemory                                  404481920
+ *     system.example                                     42
+ *     ---------- End Simulation Statistics   ----------
+ */
+class StatTester : public SimObject
+{
+  private:
+    EventFunctionWrapper event;
+
+  public:
+    StatTester(const StatTesterParams &p) :
+        SimObject(p),
+        event([this]{ setStats(); }, name() + ".event")
+    {}
+    void startup() override { schedule(event, curTick()); }
+
+  protected:
+    virtual void setStats(){}
+};
+
+class ScalarStatTester : public StatTester
+{
+
+  private:
+    ScalarStatTesterParams params;
+
+  public:
+    ScalarStatTester(const ScalarStatTesterParams &p) :
+        StatTester(p), params(p), stats(this, p) {}
+
+  protected:
+    void setStats() override;
+    struct ScalarStatTesterStats : public statistics::Group
+    {
+        ScalarStatTesterStats(
+            statistics::Group *parent,
+            const ScalarStatTesterParams &params
+        );
+        statistics::Scalar scalar;
```

Comment:
Not in this change, but in the future, does it make sense to test all of the stat types here?

Reply:
I'm not sure precisely what you mean but you can see what it looks like when I expand this in this patch where I add tests for Vector stats (yet to be incorporated PR, waiting on this to be merged): https://github.com/BobbyRBruce/gem5/commit/dff74a9950c2eae056b4ee77760c062952133cd4

Reply:
Sorry for the confusion. I understand what you did in that commit (I guess I missed the PR). That's a reasonable approach. 

I would have instead added a line here that adds a Vector stat, but your way works fine too!